### DISCUSSION TITLE
Invalid keymap argument for naked48

### DISCUSSION THREAD

**Discussion by Hororohoruru** [manna-harbour/miryoku#264] [Q&A] [answered]
Hello!

I've been using the layout since some months ago when I moved from PlanckEZ to crkbd. So far so good! Thank you for the excellent layout.

Recently, I built a naked48 keyboard and I was expecting to have a unified layout for crkbd and naked48 (and other keyboards I would use in the future), but I cannot compile the layout for the naked48. When I try to run:

```
$ qmk compile -c -kb naked48 -km miryoku -e MIRYOKU_ALPHAS=QWERTY -e MIRYOKU_EXTRA=WORKMAN -e MIRYOKU_NAV=VI
```

I get an error that says 'Invalid keymap argument'. The same command works fine for my corne, so I don't understand where the problem is coming from... here is my `custom_config.h`:

```
// Copyright 2019 Manna Harbour
// https://github.com/manna-harbour/miryoku

// This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 2 of the License, or (at your option) any later version. This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with this program. If not, see <http://www.gnu.org/licenses/>.

#pragma once

#define XXX KC_NO

#if defined (KEYBOARD_crkbd)
#define MIRYOKU_MAPPING( \
          K00,  K01,  K02,  K03,  K04,         K05,  K06,  K07,  K08,  K09, \
          K10,  K11,  K12,  K13,  K14,         K15,  K16,  K17,  K18,  K19, \
          K20,  K21,  K22,  K23,  K24,         K25,  K26,  K27,  K28,  K29, \
          N30,  N31,  K32,  K33,  K34,         K35,  K36,  K37,  N38,  N39, \
          K40,  K41\
    ) \
    LAYOUT_split_3x6_3( \
    XXX,  K00,  K01,  K02,  K03,  K04,         K05,  K06,  K07,  K08,  K09,  XXX, \
    XXX,  K10,  K11,  K12,  K13,  K14,         K15,  K16,  K17,  K18,  K19,  XXX, \
    K40,  K20,  K21,  K22,  K23,  K24,         K25,  K26,  K27,  K28,  K29,  K41, \
                      K32,  K33,  K34,         K35,  K36,  K37 \
    )
#endif

#if defined (KEYBOARD_naked48)
#define MIRYOKU_MAPPING(\
        K00,   K01,   K02,   K03,   K04,                 K05,   K06,   K07,   K08,   K09,\
        K10,   K11,   K12,   K13,   K14,                 K15,   K16,   K17,   K18,   K19,\
        K20,   K21,   K22,   K23,   K24,                 K25,   K26,   K27,   K28,   K29,\
        N30,   N31,   K32,   K33,   K34,                 K35,   K36,   K37,   N38,   N39\
        )\
        LAYOUT_ortho_4x12(\
                K00,   K01,   K02,   K03,   K04,   KC_NO, KC_NO, K05,   K06,   K07,   K08,   K09,\
                K10,   K11,   K12,   K13,   K14,   KC_NO, KC_NO, K15,   K16,   K17,   K18,   K19,\
                K20,   K21,   K22,   K23,   K24,   KC_NO, KC_NO, K25,   K26,   K27,   K28,   K29,\
                KC_NO, KC_NO, KC_NO, K32,   K33,   K34,   K35,   K36,   K37,   KC_NO, KC_NO, KC_NO\
        )
#endif
```

And then my layers are defined just below.

**Comment by manna-harbour** [comment]
I'll add support for naked48.

**Reply by Hororohoruru** [reply]
Great, thank you! As I was reading other discussions, I've seen you answering several times that the process for adding support for a new keyboard is not well documented, but that there are several committed examples. Could you link me to one of such examples? I'd love to try to do it myself if I run into this issue again in the future

**Reply by manna-harbour** [reply]
For other Miryoku implementations you can just look at the commit history for commits where boards have been added, and note the files changed. For Miryoku QMK that's a bit more complicated at the moment as commits are squashed when PR'd upstream, and branches are renamed. Instead you need to look at the old branches: https://github.com/manna-harbour/miryoku_qmk/branches/all. The dated `miryoku` branches are the old branches. Looking at the most recent branch https://github.com/manna-harbour/miryoku_qmk/tree/miryoku-20230407 you can see the commits https://github.com/manna-harbour/miryoku_qmk/commits/miryoku-20230407 and one commit adding a new keyboard https://github.com/manna-harbour/miryoku_qmk/commit/10c7ceb0fc24417526c37be71fbbeb212a46649c.

The most important file there is https://github.com/manna-harbour/miryoku_qmk/blob/10c7ceb0fc24417526c37be71fbbeb212a46649c/keyboards/takashicompany/minizone/keymaps/manna-harbour_miryoku/config.h. You need to define the layout macro. This section does not change https://github.com/manna-harbour/miryoku_qmk/blob/10c7ceb0fc24417526c37be71fbbeb212a46649c/keyboards/takashicompany/minizone/keymaps/manna-harbour_miryoku/config.h#L27-L32. This section needs to be adapted per keyboard https://github.com/manna-harbour/miryoku_qmk/blob/10c7ceb0fc24417526c37be71fbbeb212a46649c/keyboards/takashicompany/minizone/keymaps/manna-harbour_miryoku/config.h#L33-L38.

The name used here https://github.com/manna-harbour/miryoku_qmk/blob/10c7ceb0fc24417526c37be71fbbeb212a46649c/keyboards/takashicompany/minizone/keymaps/manna-harbour_miryoku/config.h#L33 comes from here https://github.com/manna-harbour/miryoku_qmk/blob/10c7ceb0fc24417526c37be71fbbeb212a46649c/keyboards/takashicompany/minizone/info.json#L12.

The key positions here https://github.com/manna-harbour/miryoku_qmk/blob/10c7ceb0fc24417526c37be71fbbeb212a46649c/keyboards/takashicompany/minizone/keymaps/manna-harbour_miryoku/config.h#L33-37 come from here https://github.com/manna-harbour/miryoku_qmk/blob/10c7ceb0fc24417526c37be71fbbeb212a46649c/keyboards/takashicompany/minizone/keymaps/default/keymap.c#L9-L12. In the case of minizone, the default keymap lacks formatting, and the default mapping has mouse buttons added, so it turned out to be not such a great example...

**Reply by Hororohoruru** [reply]
Thanks a lot for all the pointers to the code, it really helps. I managed to do it yesterday and it is working fine. For reference, I modified the layout a bit since the naked already has all rows except the bottom one separated:

```
#define MIRYOKU_MAPPING(\
        K00,   K01,   K02,   K03,   K04,                 K05,   K06,   K07,   K08,   K09,\
        K10,   K11,   K12,   K13,   K14,                 K15,   K16,   K17,   K18,   K19,\
        K20,   K21,   K22,   K23,   K24,                 K25,   K26,   K27,   K28,   K29,\
        N30,   N31,   K32,   K33,   K34,                 K35,   K36,   K37,   N38,   N39\
        )\
        LAYOUT(\
                KC_NO,  K00,    K01,    K02,   K03,   K04,                     K05,   K06,   K07,   K08,   K09,  KC_NO \
                KC_NO,  K10,    K11,    K12,   K13,   K14,                     K15,   K16,   K17,   K18,   K19,  KC_NO \
                KC_NO,  K20,    K21,    K22,   K23,   K24,                     K25,   K26,   K27,   K28,   K29,  KC_NO \
                        KC_NO,  KC_NO,  K32,   K33,   K34,   KC_NO,   KC_NO,   K35,   K36,   K37,   KC_NO, KC_NO \
        )
```

**Comment by Hororohoruru** [comment]
I tryed creating `keyboards/salicylic_acid3/naked48/keymaps/manna-harbour_miryoku` and copying `config.h` and `keymap.c` from `layouts/community/ortho_4x12/manna-harbour_miryoku` but I got an error saying 'error: implicit declaration of function LAYOUT_ortho_4x12'.

What could be the cause of this? I checked other new keyboard implementations made on the repo and I don't see anything different from what I did...

**Reply by manna-harbour** [reply]
The layout macro for this keyboard is named `LAYOUT` https://github.com/qmk/qmk_firmware/blob/0115951de12bd4c798a43a380651c34552552e85/keyboards/salicylic_acid3/naked48/rev1/info.json#L50. It doesn't use `LAYOUT_ortho_4x12`. You can see the key positions here https://github.com/qmk/qmk_firmware/blob/0115951de12bd4c798a43a380651c34552552e85/keyboards/salicylic_acid3/naked48/keymaps/default/keymap.c#L29-L35.

**Reply by manna-harbour** [reply]
Also note the naked60 mapping from Miryoku ZMK: https://github.com/manna-harbour/miryoku_zmk/blob/master/miryoku/mapping/60/naked60.h. You can you that as a reference and remove the top row. Note there there are some differences between a mapping in Miryoku QMK and Miryoku ZMK so you can't use it directly.

**Comment by manna-harbour** [comment]
Re: https://github.com/qmk/qmk_firmware/blob/69f7031682f854dc70c2a3f4ffacf82f632b6d31/users/manna-harbour_miryoku/custom_config.h.

I'd suggest adding the extra keys in the mapping only, not in each layer. Then your layers will still be compatible with all other supported keyboards. So, keep https://github.com/qmk/qmk_firmware/blob/69f7031682f854dc70c2a3f4ffacf82f632b6d31/users/manna-harbour_miryoku/custom_config.h#L21, but replace `K40`/`K41` with the actual keycodes `SC_LSPO`/`SC_RSPC`. Then remove this line https://github.com/qmk/qmk_firmware/blob/69f7031682f854dc70c2a3f4ffacf82f632b6d31/users/manna-harbour_miryoku/custom_config.h#L16 and all extra lines in the layers e.g.https://github.com/qmk/qmk_firmware/blob/69f7031682f854dc70c2a3f4ffacf82f632b6d31/users/manna-harbour_miryoku/custom_config.h#L30. If you haven't modified the layers you can remove the definitions. If necessary you can customise mappings per layer.

**Reply by Hororohoruru** [reply]
I see, thank you very much for the detailed explanation! In the end I removed it since it was just an experiment to see if I could add per-layer new keys to the mapping, thinking about using the unused keys for things I would like to have on combos until custom combos are supported. In the case of the naked48, I imagine such a modification would require to also modify the `LAYOUT` in the keyboard's `keymap/manna-harbour_miryoku` folder, right?

In any case, I understand that modifying the mappings to add new keys would require for all the compatible keyboars to accept the same number of keys... if I ever use more keys (probably getting a lumberjack at some point), I would have to have layers defined only for that keyboard inside `custom_config`.

**Reply by manna-harbour** [reply]
> In the case of the naked48, I imagine such a modification would require to also modify the `LAYOUT` in the keyboard's `keymap/manna-harbour_miryoku` folder, right?

You can do per layer custom mappings in `custom_config.h`.

> In any case, I understand that modifying the mappings to add new keys would require for all the compatible keyboars to accept the same number of keys...

If you add extra keys to the layers you also need to modify all mappings (but you can drop extra keys if not present).

> if I ever use more keys (probably getting a lumberjack at some point), I would have to have layers defined only for that keyboard inside `custom_config`.

 If you map unused keys in the mappings instead you don't have to modify the layers or modify the mappings for other keyboards at all.

**Comment by manna-harbour** [comment, marked as answer]
Added in https://github.com/manna-harbour/miryoku_qmk/commit/df1690ad257e7b0d1bf9450258f7baebfee7be9a. Here's a sample build run (with `RGBLIGHT_ENABLE=no` due to firmware space constraints on AVR): https://github.com/manna-harbour/miryoku_qmk/actions/runs/7031799035.
